### ISSUE TITLE
Bringing up all boxes in multi-box configuration fails with Ruby 3.1

### ISSUE THREAD

**Issue by ryan-ronnander** [vagrant-libvirt/vagrant-libvirt#1445]
Vagrant fails to bring up one or more boxes in a multi box format with the libvirt provider. Running multiple `vagrant up` commands will eventually bring up all vagrant boxes. The following error occurs:

```text
An error occurred while executing multiple actions in parallel.
Any errors that occurred are shown below.

An unexpected error occurred when executing the action on the
'testbox-2' machine. Please report this as a bug:

constant VagrantPlugins::ProviderLibvirt::Action::HandleBox not defined

                b2.use HandleBox
                       ^^^^^^^^^

/home/rronnander/.vagrant.d/gems/3.1.0/gems/vagrant-libvirt-0.7.0/lib/vagrant-libvirt/action.rb:38:in `block (2 levels) in action_up'
/usr/share/vagrant/gems/gems/vagrant-2.2.19/lib/vagrant/action/builtin/call.rb:47:in `call'
/usr/share/vagrant/gems/gems/vagrant-2.2.19/lib/vagrant/action/warden.rb:48:in `call'
/usr/share/vagrant/gems/gems/vagrant-2.2.19/lib/vagrant/action/builtin/box_check_outdated.rb:93:in `call'
/usr/share/vagrant/gems/gems/vagrant-2.2.19/lib/vagrant/action/warden.rb:48:in `call'
/usr/share/vagrant/gems/gems/vagrant-2.2.19/lib/vagrant/action/builtin/config_validate.rb:25:in `call'
/usr/share/vagrant/gems/gems/vagrant-2.2.19/lib/vagrant/action/warden.rb:48:in `call'
/usr/share/vagrant/gems/gems/vagrant-2.2.19/lib/vagrant/action/builder.rb:149:in `call'
/usr/share/vagrant/gems/gems/vagrant-2.2.19/lib/vagrant/action/runner.rb:89:in `block in run'
/usr/share/vagrant/gems/gems/vagrant-2.2.19/lib/vagrant/util/busy.rb:19:in `busy'
/usr/share/vagrant/gems/gems/vagrant-2.2.19/lib/vagrant/action/runner.rb:89:in `run'
/usr/share/vagrant/gems/gems/vagrant-2.2.19/lib/vagrant/machine.rb:246:in `action_raw'
/usr/share/vagrant/gems/gems/vagrant-2.2.19/lib/vagrant/machine.rb:215:in `block in action'
/usr/share/vagrant/gems/gems/vagrant-2.2.19/lib/vagrant/environment.rb:614:in `lock'
/usr/share/vagrant/gems/gems/vagrant-2.2.19/lib/vagrant/machine.rb:201:in `call'
/usr/share/vagrant/gems/gems/vagrant-2.2.19/lib/vagrant/machine.rb:201:in `action'
/usr/share/vagrant/gems/gems/vagrant-2.2.19/lib/vagrant/batch_action.rb:86:in `block (2 levels) in run'
```

This issue started occurring after a recent OpenSUSE Tumbleweed update. Occurs both with the packaged `vagrant-libvirt` provided by the distro, as well as building from source. 

**To Reproduce**
1. Initially bring up a multi-box (three box) vagrant stack with `vagrant up`.
2. Error referenced above occurs, only 2/3 of the boxes created successfully. 

**Expected behavior**

All boxes (3/3) should be created successfully with no error messages.

**Versions (please complete the following information):**:
- Libvirt version: 8.0.0
- Vagrant version [output of `vagrant version`]: 2.2.19
- Vagrant flavour [Upstream or Distro]: Distro - OpenSUSE Tumbleweed
- Vagrant plugins versions (including vagrant-libvirt) [output of `vagrant plugin list`]: 0.7.0,

**Debug Log**
Attach Output of `VAGRANT_LOG=debug vagrant up --provider=libvirt >vagrant.log 2>&1`

[vagrant.log](https://github.com/vagrant-libvirt/vagrant-libvirt/files/8005557/vagrant.log)

**A Vagrantfile to reproduce the issue:**
```ruby
Vagrant.require_version ">= 2.2.8"
Vagrant.configure("2") do |config|
  config.vm.provider "libvirt"

  VM_CPUS = 2
  VM_MEMORY = 1024
  # Second disk size in GB

  # node (max 5)
  (0..2).each do |i|
    config.vm.define "testbox-#{i}" do |node|
      config.vm.box = "generic/ubuntu2004"
      config.ssh.insert_key = false

      # Libvirt Provider
      node.vm.provider "libvirt" do |l|
        l.cpus=VM_CPUS
        l.memory=VM_MEMORY
      end
    end
  end
end

```

**Comment by ryan-ronnander** [comment]
Closing due to ruby dependencies causing this issue in Tumbleweed.

**Comment by ryan-ronnander** [comment]
Reopening this issue as it is still present with packaged versions of `vagrant-libvirt` as well as self compiled. There was a separate packaging issue after Tumbleweed moved from Ruby 3.0 -> 3.1 that caused me to close out this issue originally.

In general "actions in parallel" randomly fails for `vagrant up` (initial up or otherwise) with Ruby version 3.1. Other commands such as `halt` or `destroy` work just fine. Running `vagrant up` repeatedly (if after encountering any errors) eventually succeeds with all defined vagrant boxes up and running. 

Specific Ruby version: `ruby 3.1.0p0 (2021-12-25 revision fb4df44d16) [x86_64-linux-gnu]`

**Comment by electrofelix** [comment]
Looks like our use of autoload may be causing this

https://bugs.ruby-lang.org/issues/5653

**Comment by electrofelix** [comment]
To be more precise, autoload is still in ruby 3, however the original reason for considering deprecation of autoload was that there is an issue using it with threads, and it's likely something subtle has changed in ruby that is now causing this to be exposed

**Comment by electrofelix** [comment]
I'm hopeful that I've fixed this, but may have to reopen later if you still bump into it after the fix is released.

**Comment by tserong** [comment]
I hit this issue so backported #1451 to a fork of openSUSE Tumbleweed's vagrant-libvirt-0.7.0 (see https://build.opensuse.org/package/view_file/home:tserong:branches:Virtualization:vagrant/vagrant-libvirt/0004-Move-autoloading-to-start-of-module-1451.patch?expand=1). Note that this build is 0.7.0 plus the three patches already existing in the openSUSE build, plus my backport of #1451 - it doesn't include any other upstream commits since 0.7.0 (not sure if that's significant, but given the problem description, I assume not). Then I did some testing of both 0.7.0 and 0.7.0 plus #1451, using the following script to `vagrant up` and `vagrant destroy` 50 times, counting successes and failures:

```sh
#!/bin/bash

COUNT=0
SUCCESS=0
FAILURE=0
LOG=/tmp/vagrant-up.$$.log

echo "Testing with $(rpm -q vagrant-libvirt)"

while [ $COUNT -lt 50 ] ; do
	vagrant up --parallel >$LOG 2>&1
	if [ $? -eq 0 ] ; then
		grep 'booted and ready' $LOG
		let SUCCESS=$SUCCESS+1
	else
		grep 'b2\.use' $LOG
		let FAILURE=$FAILURE+1
	fi
	echo "SUCCESS: $SUCCESS, FAILURE: $FAILURE"
	sleep 2
	vagrant destroy --force >/dev/null 2>&1
	sleep 2
	let COUNT=$COUNT+1
done
```

Alas, the change doesn't seem to have had any real effect, it still fails intermittently. The regular 0.7.0 vagrant-libvirt build gave me "SUCCESS: 21, FAILURE: 29", and my patched version gave "SUCCESS: 19, FAILURE: 31".

The failures were mostly to do with `HandleBox` and `WaitForCommunicator`, plus a handful of `SyncedFolders`.

```
> grep 'b2\.use' stock.log|sort|uniq -c
     29                 b2.use HandleBox
      5                 b2.use SyncedFolders
     14                 b2.use WaitForCommunicator, [:running]

> grep 'b2\.use' backport-1451.log|sort|uniq -c
     31                 b2.use HandleBox
      7                 b2.use SyncedFolders
     17                 b2.use WaitForCommunicator, [:running]
```

**Comment by electrofelix** [comment]
I guess it's still present though I don't know what will be the fix since it appears to be an issue with use of supported ruby behaviour.

**Comment by electrofelix** [comment]
One thing I've overlooked is that HandleBox is looking to autoload from vagrant itself rather than a local autoload. So I wonder if there is some quirk occuring there.

One option is to move all of the autoloading to a method that does a require of the necessary modules instead and insure that is called at the start of each action function to avoid adding the overhead of requiring the files if the main action isn't being called. Would require updating the constants being referenced to use the full module/class names but that might be worth exploring. Not necessarily the right approach but might be a good enough fix for now until the autoload behaviour is better understood.

Might be a while before I can properly look at this, but if someone else has a bit of time it might be worth exploring what happens if you implement the above change.

**Comment by tserong** [comment]
I did a little experiment based on your above comment @electrofelix. I didn't make a separate method to require everything (because TBH I wasn't sure how best to implement that - it's been a while since I did any ruby), I just removed the `autoload` lines for each of `:HandleBox`, `:SyncedFolders` and `:WaitForCommunicator`, then added matching `requires` before the various `b*.use` statements. Rationale: those were the three classes that I kept seeing the errors on:

```diff --git a/lib/vagrant-libvirt/action.rb b/lib/vagrant-libvirt/action.rb
index e8bc1ec..b186e51 100644
--- a/lib/vagrant-libvirt/action.rb
+++ b/lib/vagrant-libvirt/action.rb
@@ -56,11 +56,8 @@ module VagrantPlugins
       autoload :PackageSetupFiles, 'vagrant/action/general/package_setup_files'
       autoload :PackageSetupFolders, 'vagrant/action/general/package_setup_folders'
       autoload :SSHRun, 'vagrant/action/builtin/ssh_run'
-      autoload :HandleBox, 'vagrant/action/builtin/handle_box'
-      autoload :SyncedFolders, 'vagrant/action/builtin/synced_folders'
       autoload :SyncedFolderCleanup, 'vagrant/action/builtin/synced_folder_cleanup'
       autoload :ProvisionerCleanup, 'vagrant/action/builtin/provisioner_cleanup'
-      autoload :WaitForCommunicator, 'vagrant/action/builtin/wait_for_communicator'
 
       # Include the built-in & general modules so we can use them as top-level things.
       include Vagrant::Action::Builtin
@@ -91,6 +88,7 @@ module VagrantPlugins
                 b2.use StartDomain
               else
                 b2.use HandleStoragePool
+                require 'vagrant/action/builtin/handle_box'
                 b2.use HandleBox
                 b2.use HandleBoxImage
                 b2.use CreateDomainVolume
@@ -99,6 +97,7 @@ module VagrantPlugins
                 b2.use Provision
                 b2.use PrepareNFSValidIds
                 b2.use SyncedFolderCleanup
+                require 'vagrant/action/builtin/synced_folders'
                 b2.use SyncedFolders
                 b2.use PrepareNFSSettings
                 b2.use ShareFolders
@@ -108,6 +107,7 @@ module VagrantPlugins
 
                 b2.use StartDomain
                 b2.use WaitTillUp
+                require 'vagrant/action/builtin/wait_for_communicator'
                 b2.use WaitForCommunicator, [:running]
 
                 b2.use ForwardPorts
@@ -159,6 +159,7 @@ module VagrantPlugins
 
                 b3.use PrepareNFSValidIds
                 b3.use SyncedFolderCleanup
+                require 'vagrant/action/builtin/synced_folders'
                 b3.use SyncedFolders
 
                 # Start it..
@@ -167,6 +168,7 @@ module VagrantPlugins
                 # Machine should gain IP address when comming up,
                 # so wait for dhcp lease and store IP into machines data_dir.
                 b3.use WaitTillUp
+                require 'vagrant/action/builtin/wait_for_communicator'
                 b3.use WaitForCommunicator, [:running]
 
                 b3.use ForwardPorts
```

With that change, I ran my script to `vagrant up` and `vagrant destroy` 50 times, and it succeeded every single time. This suggests to me that this approach may be viable.

**Comment by mattiasb** [comment]
I'm using this workaround while I'm waiting for Fedora to get `vagrant-libvirt` upgraded to incorporate this fix ([RHBZ#2132090](https://bugzilla.redhat.com/show_bug.cgi?id=2132090)):
```shell

function vagrant-up {
    local boxes

    boxes="$( vagrant status --machine-readable                                \
            | cut -d "," -f 2                                                  \
            | tr -s '\n' '\n'                                                  \
            | uniq -u                                                          \
            )"

    for box in $boxes; do
        vagrant up "${box}"
    done
}

```
